src/storage/src/source/postgres/replication.rs

```diff
@@ -500,68 +508,48 @@ pub(crate) fn render<G: Scope<Timestamp = MzOffset>>(
-                                    Err(PostgresError::PublicationMissing(publication)) => {
-                                        let err = DefiniteError::PublicationDropped(publication);
-                                        for (oid, outputs) in table_info.iter() {
-                                            for output_index in outputs.keys() {
-                                                let update = (
-                                                    (
-                                                        *oid,
-                                                        *output_index,
-                                                        Err(DataflowError::from(err.clone())),
-                                                    ),
-                                                    data_cap_set[0].time().clone(),
-                                                    1,
-                                                );
-                                                data_output.give_fueled(&data_cap_set[0], update).await;
-                                            }
-                                        }
-                                        definite_error_handle.give(
-                                            &definite_error_cap_set[0],
-                                            ReplicationError::Definite(Rc::new(err)),
-                                        );
-                                        return Ok(());
-                                    }
-                                    Err(e) => Err(TransientError::from(e))?,
-                                }
-                            };
-                            for (&oid, outputs) in table_info.iter() {
-                                for (output_index, info) in outputs {
-                                    if errored.contains(output_index) {
-                                        trace!(%id, "timely-{worker_id} output index {output_index} \
-                                            for oid {oid} skipped");
-                                        continue;
-                                    }
-                                    match verify_schema(oid, &info.desc, &upstream_info, &*info.casts) {
-                                        Ok(()) => {
-                                            trace!(%id, "timely-{worker_id} schema of output \
-                                                index {output_index} for oid {oid} valid");
-                                        }
-                                        Err(err) => {
-                                            trace!(%id, "timely-{worker_id} schema of output \
-                                                index {output_index} for oid {oid} invalid");
+
+                        // Take the opportunity to report any schema validation errors.
+                        while let Ok(error) = schema_errors.try_recv() {
```

Comment:
Using the tokio `Watch` here is just as good since we only ever want to process the most recent schema

Reply:
I tried this out: A `Watch` is a bit annoying to use in the schema validator task because we can have multiple errors, so we need to collect them all in a vec and then send them at the end. In the replication reader we also can't `take` the errors because we only get immutable references to them, so we have to clone them and then remember that we've already seen them. Both doable but more noisy than using a channel.

We could use an `Arc<Mutex<...>` instead of a `Watch` to solve the immutable references issue, but we'd still have to do the collection into a vec in the schema validator task.